client/engine/chainservice/adjudicator/reclaim.go

```diff
@@ -0,0 +1,59 @@
+package NitroAdjudicator
+
+import (
+	"fmt"
+
+	"github.com/statechannels/go-nitro/channel/state/outcome"
+)
+
+// computeReclaimEffects mirrors on chain code.
+// It computes side effects for the reclaim function. Returns updated allocations for the source, computed by finding the guarantee in the source for the target, and moving money out of the guarantee and back into the ledger channel as regular allocations for the participants.
+func computeReclaimEffects(sourceAllocations []outcome.Allocation, targetAllocations []outcome.Allocation, indexOfTargetInSource uint) ([]outcome.Allocation, error) {
+
+	newSourceAllocations := make([]outcome.Allocation, len(sourceAllocations)-1)
+	guarantee := sourceAllocations[indexOfTargetInSource]
+	guaranteeData, err := outcome.DecodeIntoGuaranteeMetadata(guarantee.Metadata)
+
+	if err != nil {
+		return []outcome.Allocation{}, err
+	}
+
+	var (
+		foundTarget, foundLeft, foundRight bool
+	)
+
+	k := 0
+	for i := 0; i < len(sourceAllocations); i++ {
+		if i == int(indexOfTargetInSource) {
+			foundTarget = true
+			continue
+		}
+		newSourceAllocations[k] = outcome.Allocation{
+			Destination:    sourceAllocations[i].Destination,
+			Amount:         sourceAllocations[i].Amount,
+			AllocationType: sourceAllocations[i].AllocationType,
+			Metadata:       sourceAllocations[i].Metadata,
+		}
+		if sourceAllocations[i].Destination == guaranteeData.Left {
+			newSourceAllocations[k].Amount.Add(newSourceAllocations[k].Amount, targetAllocations[0].Amount)
```

Comment:
Does this work if destinations are repeated in `sourceAllocation`?

Reply:
If destinations are repeated, this could lead to a problem here (multiple entries will be incremented). 
We can 
1. add client checks about not having repeated destinations
2. add a `continue` statement on L39 and another on L43
3. revert the transaction if the total amount reclaimed is too high (see your other comment thread)

Reply:
2. EDIT `continue` statements won't do it. But we could check `!foundLeft` on L37 etc.